File: lib/src/parser/beam_parser.dart
import 'package:musicxml_parser/src/models/beam.dart';
import 'package:xml/xml.dart';

class BeamParser {
  /// Parses beam elements from a note element.
  ///
  /// [noteElement] - The XML element representing the note.
  /// [noteIndex] - The index of the note within the measure.
  /// [measureNumber] - The measure number this beam belongs to.
  static List<Beam> parse(
      XmlElement noteElement, int noteIndex, String measureNumber) {
    final beams = <Beam>[];

    final beamElements = noteElement.findElements('beam');
    for (final beamElement in beamElements) {
      final beamType = beamElement.innerText.trim();
      final beamNumberAttr = beamElement.getAttribute('number');
      final beamNumber =
          beamNumberAttr != null ? int.tryParse(beamNumberAttr) ?? 1 : 1;

      beams.add(Beam(
        number: beamNumber,
        type: beamType,
        measureNumber: measureNumber,
        noteIndices: [noteIndex], // 后续再合并
      ));
    }

    return beams;
  }

  /// 将单个 note 的 beam 元素合并为完整的 beam 组
  ///
  /// 这个方法会基于 beam 的 number 和 type (begin, continue, end) 属性
  /// 将相关的 beam 元素连接起来，形成完整的 beam 组
  static List<Beam> mergeBeams(List<Beam> individualBeams, String measureNumber) {
    if (individualBeams.isEmpty) return [];

    // 按 beam number 分组
    final beamsByNumber = <int, List<Beam>>{};

    for (final beam in individualBeams) {
      beamsByNumber.putIfAbsent(beam.number, () => []).add(beam);
    }

    final result = <Beam>[];

    // 对每个 beam number 分别处理
    beamsByNumber.forEach((number, beams) {
      // 对 beams 按照 noteIndex 排序，确保顺序正确
      beams.sort((a, b) => a.noteIndices.first.compareTo(b.noteIndices.first));

      // 临时存储当前处理的 beam 组
      Beam? currentBeam;
      List<int> currentNoteIndices = [];
      String? currentType;

      // 按顺序处理每个 beam
      for (final beam in beams) {
        // 如果当前 beam 是 begin 或前一个 beam 组已完成，开始新的 beam 组
        if (beam.type == 'begin' || currentBeam == null) {
          // 如果有未完成的 beam 组，先加入结果
          if (currentBeam != null && currentNoteIndices.isNotEmpty) {
            result.add(Beam(
              number: currentBeam.number,
              type: currentType ?? 'unknown',
              measureNumber: measureNumber,
              noteIndices: List.from(currentNoteIndices),
            ));
          }

          // 开始新的 beam 组
          currentBeam = beam;
          currentType = beam.type;
          currentNoteIndices = List.from(beam.noteIndices);
        }
        // 如果是 continue 或 end，添加到当前 beam 组
        else if (beam.type == 'continue' || beam.type == 'end') {
          currentNoteIndices.addAll(beam.noteIndices);

          // 如果是 end，完成当前 beam 组
          if (beam.type == 'end') {
            result.add(Beam(
              number: currentBeam.number,
              type: currentBeam.type, // 保持原始的 begin 类型
              measureNumber: measureNumber,
              noteIndices: List.from(currentNoteIndices),
            ));

            // 重置状态
            currentBeam = null;
            currentNoteIndices = [];
            currentType = null;
          }
        }
        // 对于单独的 forward hook 或 backward hook
        else if (beam.type == 'forward hook' || beam.type == 'backward hook') {
          // 这些通常是独立的，直接添加到结果
          result.add(beam);
        }
      }

      // 处理可能剩余的未完成 beam 组
      if (currentBeam != null && currentNoteIndices.isNotEmpty) {
        result.add(Beam(
          number: currentBeam.number,
          type: currentType ?? 'unknown',
          measureNumber: measureNumber,
          noteIndices: List.from(currentNoteIndices),
        ));
      }
    });

    return result;
  }
}

File: MusicXMLParser/Parser/BeamParser.cs
// Assuming the necessary using statements for MusicXML models
// e.g., using MusicXMLParser.Models;
using System.Xml.Linq;
using System.Collections.Generic;
using System.Linq;
using MusicXMLParser.Models; // Assuming this is where Beam model is

namespace MusicXMLParser.Parser
{
    public static class BeamParser
    {
        /// <summary>
        /// Parses beam elements from a note element.
        /// </summary>
        /// <param name="noteElement">The XML element representing the note.</param>
        /// <param name="noteIndex">The index of the note within the measure.</param>
        /// <param name="measureNumber">The measure number this beam belongs to.</param>
        /// <returns>A list of Beam objects parsed from the note element.</returns>
        public static List<Beam> Parse(
            XElement noteElement, int noteIndex, string measureNumber)
        {
            var beams = new List<Beam>();

            var beamElements = noteElement.Elements("beam");
            foreach (var beamElement in beamElements)
            {
                var beamType = beamElement.Value.Trim();
                var beamNumberAttr = beamElement.Attribute("number")?.Value;
                var beamNumber = !string.IsNullOrEmpty(beamNumberAttr) && int.TryParse(beamNumberAttr, out int num) ? num : 1;

                beams.Add(new Beam(
                    number: beamNumber,
                    type: beamType,
                    measureNumber: measureNumber,
                    noteIndices: new List<int> { noteIndex } // Will be merged later
                ));
            }

            return beams;
        }

        /// <summary>
        /// Merges individual note beam elements into complete beam groups.
        /// This method connects related beam elements based on their number and type (begin, continue, end)
        /// attributes to form complete beam groups.
        /// </summary>
        /// <param name="individualBeams">A list of beams, each associated with a single note.</param>
        /// <param name="measureNumber">The measure number these beams belong to.</param>
        /// <returns>A list of merged Beam objects representing complete beam groups.</returns>
        public static List<Beam> MergeBeams(
            List<Beam> individualBeams, string measureNumber)
        {
            if (individualBeams == null || !individualBeams.Any()) return new List<Beam>();

            // Group beams by beam number
            var beamsByNumber = individualBeams
                .GroupBy(b => b.Number)
                .ToDictionary(g => g.Key, g => g.ToList());

            var result = new List<Beam>();

            // Process each beam number separately
            foreach (var entry in beamsByNumber)
            {
                var number = entry.Key;
                var beams = entry.Value;

                // Sort beams by noteIndex to ensure correct order
                beams.Sort((a, b) => a.NoteIndices.First().CompareTo(b.NoteIndices.First()));

                // Temporary storage for the currently processed beam group
                Beam? currentBeamContext = null;
                List<int> currentNoteIndices = new List<int>();
                string? currentType = null; // Stores the type of the start of the beam ('begin')

                // Process each beam in order
                foreach (var beam in beams)
                {
                    // If current beam is 'begin' or the previous beam group is completed, start a new beam group
                    if (beam.Type == "begin" || currentBeamContext == null)
                    {
                        // If there's an unfinished beam group, add it to the result
                        if (currentBeamContext != null && currentNoteIndices.Any())
                        {
                            result.Add(new Beam(
                                number: currentBeamContext.Number,
                                type: currentType ?? "unknown", // Should be the type of the first beam in the group
                                measureNumber: measureNumber,
                                noteIndices: new List<int>(currentNoteIndices)
                            ));
                        }

                        // Start a new beam group
                        currentBeamContext = beam;
                        currentType = beam.Type; // This is the 'begin' type
                        currentNoteIndices = new List<int>(beam.NoteIndices);
                    }
                    // If it's 'continue' or 'end', add to the current beam group
                    else if (beam.Type == "continue" || beam.Type == "end")
                    {
                        if (currentBeamContext != null) // Ensure there's an active beam group
                        {
                            currentNoteIndices.AddRange(beam.NoteIndices);

                            // If it's 'end', complete the current beam group
                            if (beam.Type == "end")
                            {
                                result.Add(new Beam(
                                    number: currentBeamContext.Number,
                                    type: currentType, // Use the stored 'begin' type
                                    measureNumber: measureNumber,
                                    noteIndices: new List<int>(currentNoteIndices)
                                ));

                                // Reset state
                                currentBeamContext = null;
                                currentNoteIndices = new List<int>();
                                currentType = null;
                            }
                        }
                        else
                        {
                            // This case (continue/end without a begin) might indicate a malformed MusicXML
                            // or a beam starting mid-measure without a prior 'begin' in this segment.
                            // Depending on strictness, this could be an error or handled gracefully.
                            // For now, we'll create a new beam segment.
                             result.Add(new Beam(
                                number: beam.Number,
                                type: beam.Type, // Could be 'continue' or 'end'
                                measureNumber: measureNumber,
                                noteIndices: new List<int>(beam.NoteIndices)
                            ));
                        }
                    }
                    // For standalone 'forward hook' or 'backward hook'
                    else if (beam.Type == "forward hook" || beam.Type == "backward hook")
                    {
                        // These are typically independent and added directly to the result
                        result.Add(beam);
                    }
                }

                // Handle any remaining unfinished beam group
                if (currentBeamContext != null && currentNoteIndices.Any())
                {
                    result.Add(new Beam(
                        number: currentBeamContext.Number,
                        type: currentType ?? "unknown",
                        measureNumber: measureNumber,
                        noteIndices: new List<int>(currentNoteIndices)
                    ));
                }
            }
            return result;
        }
    }
}
